File: January 2025/25.cpp
#include <iostream>
using namespace std;
class Solution {
  public:
    Node* findFirstNode(Node* head) {
        // Initialize slow and fast pointers
        Node* slow = head;
        Node* fast = head;
        
        // Detect if there is a loop using Floyd's Cycle-Finding Algorithm
        while (fast != NULL && fast->next != NULL) {
            slow = slow->next;
            fast = fast->next->next;
            if (slow == fast) {
                break;
            }
        }
        
        // If no loop is found, return NULL
        if (fast == NULL || fast->next == NULL) {
            return NULL;
        }
        
        // Find the start of the loop
        slow = head;
        while (slow != fast) {
            slow = slow->next;
            fast = fast->next;
        }
        
        return slow; // Return the node itself
    }
};

File: January 2025/25.py
# Node Class
class Node:
    def __init__(self, data):   # data -> value stored in node
        self.data = data
        self.next = None

class Solution:
    def findFirstNode(self, head):
        # Initialize slow and fast pointers
        slow = head
        fast = head
        
        # Detect if there is a loop using Floyd's Cycle-Finding Algorithm
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if slow == fast:
                break
        
        # If no loop is found, return None
        if not fast or not fast.next:
            return None
        
        # Find the start of the loop
        slow = head
        while slow != fast:
            slow = slow.next
            fast = fast.next
        
        return slow  # Return the node itself
